training/MESSAGING_OVERVIEW.md

```diff
@@ -0,0 +1,138 @@
+
+<a id="toc"></a>
+## Table of Contents
+1. [Overview](#msgOverview)
+2. [Sending a Message to the Server](#sendReq)
+3. [Recieving a Request on the Server](#receiveReq)
+4. [Creating a new Chapel Module](#newModule)
+5. [Try-It Example](#tryIt)
+
+<a id="msgOverview"></a>
+## Overview
+
+Arkouda uses ZeroMQ to handle messaging between our Python client and Chapel Server. This communication enables Arkouda to process large scale python data frameworks efficiently by leveraging Chapel's compiled and parallel nature. 
+
+To accomplish this, we use a command map structure in Chapel to map a string name to a Chapel function within a server Module. Adding a new module is simple. At the root level there is a `ServerModules.cfg` file that lists all Chapel Modules available for Command Map arguments. To add a new module, simply add the name of the module to this file. To learn how to structure a new Chapel Module, go to [Creating a New Module](#newModule).
+
+<a id="sendReq"></a>
+## How to Send a Request to the Server
+
+In general, a server request will look something like this:
+
+`repMsg = cast(str, generic_msg(cmd="serverFunction", args={"arg1": arg1, "arg2": arg2}))`
+
+We cast the return from `generic_msg` to a `str` in most cases to ensure our return is in a string format. The alternative is `memoryview`, which is a binary representation.
+
+To break this request down, `cmd` is the string name that maps to the function you want to use in Chapel. In most cases the string name will be almost identical to the Chapel function name to keep things easy to maintain. `args` is a JSON of all the arguments being passed to the server for processing. 
+
+<a id="receiveReq"></a>
+## Receiving a Request on the Server
+
+As mentioned above, the Command Map handles mapping the given `cmd` string to the proper function in the Chapel based on the available Chapel Modules. In order for a function within a module to be found by the Command Map it must first be registered. An example of this, using the request in the previous section as an example, would be as follows:
+
+`registerFunction("serverFunction", serverFunctionMsg, getModuleName());`
+
+The three components of the `registerFunction` method are the `cmd` string, the Chapel function name, and the Module name which in most cases we use `getModuleName()` to get the name of the module the function is called from.
+
+All messages sent to the server are in JSON format. So before any arguments sent from the client can be used, they must first be parsed. To do this use the function `parseMessageArgs(payload, argSize)`. This will parse your message arguments into a Map that can then be accessed using accessor methods. The most common of these accessor methods is `getValueOf(argName)` which will return the value of a argument that matches the passed in name.

```

Comment:
Between these sections we need to go into details on what the expected format for `serverFunctionMsg` is, `proc serverFunctionMsg(cmd: string, payload: string, argSize: int, st: borrowed SymTab): MsgTuple throws {}`. We should highlight, that the server will not compile if it does not match that format and what each parameter is for.

Reply:
Also think it might be nice list some of the other common accessor methods `getIntValue`, `getBoolValue`, ... since these have a slightly different format and will save people from doing the casting themselves